### ISSUE TITLE
[META]Allow specifying different and multiple fleet servers in agent policy

### ISSUE THREAD

**Issue by ruflin** [elastic/fleet-server#903]
Over the past few weeks we have seen several users trying to run multiple fleet-servers. This issue is to discuss the different scenarios and share an initial thoughts on how we could approach this. The goal is to come a conclusion and feed the information in the documentation and have it guideline for future features.

# Core concepts

* No load balancing: It is not the Elastic Agent jobs to load balance between multiple fleet-servers. This must be done at the infrastructure level through proxy or DNS.
* Fail over: An Elastic Agent supports multiple fleet-server urls as failover. By default the first url is picked.
* All info is shared across fleet-servers: Each fleet-server has always the same information no matter where it is deployed.

# Scenarios

Below is a list of scenarios which should describe the expected behaviours to follow the above core concepts. Not all scenarios are supported today.

## Scenario 1: Multiple fleet-servers, all Elastic Agent connect to all fleet-servers

In scenario 1 the users had multiple fleet-servers for redundancy or scale purpose. The user is expected to setup a proxy in front of the fleet-servers or use DNS to access the multiple fleet-servers. In the Fleet UI, a single fleet-server url is used.

## Scenario 2: Elastic Cloud only

The user connects all its Elastic Agent to the fleet-server in Elastic Cloud. As ESS already has a proxy in front and allows to spin up redundant fleet-servers, the setup is already as expected today.

## Scenario 3: Elastic Cloud fleet-server and on prem

The users is using Elastic Cloud with the fleet-server but also runs its on prem fleet-server to have the fleet-server closer to its Elastic Agents. By default the user wants to have its local fleet-servers to be used. In the Fleet UI, the user puts and additional fleet-server url before the Elastic Cloud fleet-server url. The user defined URL is the one used by default. In case the local fleet-server url is not reachable, Elastic Agents fall back to the Elastic Cloud fleet-server url.

The local fleet-server is expected to have the version in sync with the hosted fleet-server.

## Scenario 4: Multi policy with multi data center

In this scenario the user has multiple data centers with local fleet-servers and specific policies to each data center. In the Fleet UI, a global fleet-server is specified and in addition a fleet-server per policy can be specified. The fleet-server specified in the policy is the first in the list so it will be picked as the default for all Elastic Agent which are part of the policy.

**Comment by ruflin** [comment]
@scunningham @joshdover @michel-laterman @nimarezainia  Would be great to get your take on the above.

**Comment by scunningham** [comment]
@ruflin Makes sense.  We could consider an automatic mode where the agent would regularly ping each listed URL and determine shortest route either by round trip time or trace-route.  However, that seems like a nice to have and not sure customers would use it.

I think the bigger concern is going to be the multitude of outbound connections from the integrations to elastic.  There is very little volume to the Fleet Servers in comparison. Is the idea here to allow different ES outputs per policy?

**Comment by ruflin** [comment]
@scunningham To keep the discussion focused on fleet-server only, should we open a different issue / thread our the ES outputs per policy to set some guidelines there too?

**Comment by nimarezainia** [comment]
thanks @ruflin for this discussion. I think if we work towards defining what Scenario 4 would look like - i.e FS per policy - it would address many of the multi-site use cases we see today. In the near future with either Logstash support or Secure proxy concept, we could move towards having control and data plane aggregated at a site and reduce the number of connections coming out of some of these DC or DMZ sites.

What should be the next steps? I will clean up the multi-site document to be more generic and we can use that.

**Comment by SHolzhauer** [comment]
Hi, wanted to pitch in regarding scenario 4.

A use case we are working/struggling with is not necessarily multiple data centers but different  connection types.

There is one Elastic Stack running on premise (cloud but self managed) where the bulk of the servers are using internal
dns/connections to connect to a loadbalancer with 2+ (HA) fleet servers.

There are also endpoints (laptops) for the employees with agent installed. These laptops have an always on VPN allowing them to connect to the internal dns/connections (the same as the servers). However when out of office, disabled VPN or when you isolate a host. The internal connections are unavailable.

So it's a different setup, but allowing for endpoints to be specified based on agent policy should resolve the issues/setup complications.

**Comment by ruflin** [comment]
@SHolzhauer Thanks for chiming in. When the VPN is disabled, what fleet-server do you expect these Elastic Agents to connect to? Is it a global fallback one?

@nimarezainia I think first, the teams (Elastic Agent / Fleet / Security)  need to agree that above scenarios is what we expect. Especially 4 I think needs a bit more discussions as it impacts also Fleet. But if everyone is aligned, we can get it moving.

**Comment by michel-laterman** [comment]
So my understanding of the proposals is that if we specify more than one server address, the agent will try them in order so that the 1st address (either from the settings or from the policy) will be attempted before any others (such as the cloud address if available)?

**Comment by ruflin** [comment]
@michel-laterman Correct. First one wins as long as it is reachable.

**Comment by SHolzhauer** [comment]
>@SHolzhauer Thanks for chiming in. When the VPN is disabled, what fleet-server do you expect these Elastic Agents to connect to? Is it a global fallback one?

@ruflin When the VPN is **enabled** we would expect it to use the internal fleet endpoint.  
When the vpn is **disabled** or when the host is **isolated** we would expect it to use the external fleet endpoint.

But if we can specify a fleet and ES endpoint per agent policy we already have a solution.

**Comment by aarju** [comment]
Here is a real world 'multi cloud' scenario we are looking at internally in Elastic.  We want to migrate to using Fleet for security and o11y in Elastic Cloud. Elastic Cloud currently spans 53 different cloud regions across multiple providers. Each of these regions has it's own stack that we collect the o11y and security data. We then use Cross Cluster Search so we can use a single 'overview' cluster to query and alert on data across all of the regional clusters. For a deep dive of our current configuration you can check out [this blog post](https://www.elastic.co/blog/use-cross-cluster-search-elastic-cloud-observability). 

The question is how to do we migrate from the current filebeat and auditbeat system to using Fleet to deploy and manage elastic agents in a multi-region multi-cloud environment? Instead of having 53 different isolated Fleet servers we would like to have a single 'primary' fleet server that can manage the policies, agents, and integrations for all of the other fleet servers. Without that if we need to use an integration such as OSQuery or Elastic Security's Host Isolation we have to log into each of the 53 fleet servers to run the query.

**Comment by aarju** [comment]
Thanks for the input @mostlyjason!

We are really looking forward to this capability in 8.2, I think that will help a lot with this problem.
>Currently, the control plane and data plane must use the same cluster. However, we are planning to add Logstash output support in 8.2 https://github.com/elastic/kibana/issues/104987, and remote Elasticsearch output shortly after. That means you could have a dedicated cluster for your control plane in Fleet, and ship the data to your regional clusters. You could also use CCS across that data.

**Comment by ruflin** [comment]
Assigning this to @ph to take over the lead on it. I would still like to see us driving this to a conclusion on the long term vision / goal to guide us on any future implementations. @joshdover

**Comment by joshdover** [comment]
It seems there general preference for exploring option 4, being able to assign a Fleet Server host(s) per Agent policy, which generally fits well with how we expect users to use agent policies to group agents. IMO it makes sense to start there and see which use cases we can't solve with this solution.

**Comment by joshdover** [comment]
Think I jumped the gun a bit on this one. Looking closer at https://github.com/elastic/beats/blob/b5e94143d774f2434432a578f7ef5bbd71002bca/x-pack/elastic-agent/pkg/remote/client.go#L241, it appears that we do currently round-robin Fleet Servers and choose the host that was used least recently. We would need to stop doing that before exploring option 4.

**Comment by nimarezainia** [comment]
> Think I jumped the gun a bit on this one. Looking closer at https://github.com/elastic/elastic-agent/blob/ce95d6b5f36a43a927112517e91885570305e219/internal/pkg/remote/client.go#L222, it appears that we do currently round-robin Fleet Servers and choose the host that was used least recently. We would need to stop doing that before exploring option 4.

@joshdover 
if the fleet server list is per policy, the behavior you describe here would pertain to those in that policy correct?

**Comment by scunningham** [comment]
I suppose that's another way to do it.  We did a similar design at Endgame where Agents could be automatically assigned to a specific policy by applying a prioritized set of rules to metadata regularly.  We did, however, also have the concept of statically assigned agents; those for which the policy would never change.   This was desirable for certain classes of user personas, for example C level executives.  In those cases you may still need a mechanism to prioritize amongst a set of fleet servers.

Even within a policy, it may be the case that not all fleet servers are treated equally:  for example; I want the customer to try to hit the VPN fleet server first and if not accessible only then fallback to a regional or global fleet server.

**Comment by joshdover** [comment]
I personally prefer we decouple the dynamic policy assignment feature from having a prioritized Fleet Server hosts list. The latter is generically useful for more simple deployments that don’t need dynamic policy assignments and for the case Sean mentioned above. A prioritized list is also likely much lower effort than dynamic policy assignment to deliver and would unblock being able to release documentation/guidance on this.

**Comment by nchaulet** [comment]
@michel-laterman I am wondering if there is any Fleet server to do here, or it's just a matter of populating differently `fleet.hosts` in the agent policy from Kibana depending on what the user is configuring, and showing different enroll command too.

**Comment by jlind23** [comment]
@jen-huang Nima is assigned to that issue, shouldn't it be an engineer from your team instead?

**Comment by jlind23** [comment]
Discussing this with @blakerouse and @michalpristas today and no changes are needed on fleet-server's end.

**Comment by nimarezainia** [comment]
https://github.com/elastic/kibana/issues/137785
